py/redrock/rebin.py

```diff
@@ -125,24 +288,190 @@ def trapz_rebin(x, y, xnew=None, edges=None):
 
     return result
 
+def trapz_rebin_batch_gpu(x, y, xnew=None, edges=None, myz=None):
+    """Rebin y(x) flux density using trapezoidal integration between bin edges
+    GPU algorithm rebins in batch for all redshifts and bases and returns
+    3d array.
+
+    Notes:
+        y is interpreted as a density, as is the output, e.g.
+
+        >>> x = np.arange(10)
+        >>> y = np.ones(10)
+        >>> trapz_rebin(x, y, edges=[0,2,4,6,8])  #- density still 1, not 2
+        array([ 1.,  1.,  1.,  1.])
+
+    Args:
+        x (array): input x values.
+        y (array): input y values (for all bases).
+        edges (array): (optional) new bin edges.
+        myz (array): redshifts
+
+    Returns:
+        cp.array: integrated results with shape (nz, nbin, nbasis)
+            where nbin = len(results) = len(edges)-1
+
+    Raises:
+        ValueError: if edges are outside the range of x or if len(x) != len(y)
+
+    """
+    #Use global to import cupy here - it will not be needed to be imported
+    #in any other method
+    global cp
+    import cupy as cp
+
+    if edges is None:
+        edges = centers2edges(xnew)
+    else:
+        edges = np.asarray(edges)
+    edges = cp.array(edges, dtype=cp.float64)
+
+    if (myz is None):
+        myz = cp.array([0], dtype=cp.float64)
+    myz = cp.array(myz, dtype=cp.float64)
+
+    x = cp.array(x, dtype=cp.float64)
+    y = cp.array(y, dtype=cp.float64)
+
+    #Must multiply x by 1+z for comparison, only need to look at max/min cases
+    if (edges[0] < x[0]*(1+myz.max()) or edges[-1] > x[-1]*(1+myz.min())):
+        raise ValueError('edges must be within input x range')
+
+    if (not np.allclose(x[-1]-x[-2], x[1]-x[0])):
+        #Template wavelengths are unevenly spaced - use special kernel
+        return trapz_rebin_batch_gpu_unevenly_spaced(x, y, edges, myz)
+
+    if (len(y.shape) == 2):
+        nbasis = y.shape[0]
+    else:
+        nbasis = 1
 
-def rebin_template(template, z, dwave):
+    # Load CUDA kernel
+    cp_module = cp.RawModule(code=cuda_source)
+    batch_trapz_rebin_kernel = cp_module.get_function('batch_trapz_rebin')
+
+    nbin = cp.int32(len(edges)-1)
+    nz = cp.int32(len(myz))
+    n = nbin*nbasis*nz
+    nt = cp.int32(len(x))
+
+    blocks = (n+block_size-1)//block_size
+    result = cp.empty((nz, nbin, nbasis), dtype=cp.float64)
+
+    #Launch kernel and syncrhronize
+    batch_trapz_rebin_kernel((blocks,), (block_size,), (x, y, edges, myz, result, nz, nbin, nbasis, nt))
+    #cp.cuda.Stream.null.synchronize()
+
+    #Squeeze array and remove dims of 1
+    if (nz == 1 or nbasis == 1):
+        result = result.squeeze()
+    return result
+
+def trapz_rebin_batch_gpu_unevenly_spaced(x, y, edges, myz):
```

Comment:
This issue will go away since it looks like we can just use `trapz_rebin_batch_gpu_unevenly_spaced` without losing speed.  I had developed the other method first before I realized that the QSO templates were unevenly spaced in wavelength.  This will simplify some things.

- For the single user-facing method possibility, would it be ok to pass `myz` with default arg None to that method, slightly changing its existing signature?  If so then I think it is viable but would probably still need to propagate to the current method (branching to that instead of the numba method).